### DISCUSSION TITLE
Locked folders

### DISCUSSION THREAD

**Discussion by justjakka** [slskd/slskd#647] [Ideas]
Would be nice to have the ability to set up locked folders with user group management

**Comment by jpdillingham** [comment]
Thanks for the suggestion, but this is very unlikely to be added.

I added the ability to limit speeds per user category so that instead of locking files and making them completely unobtainable, you can instead put them behind a very slow connection and limit all of the users in a given group to single file downloads.  The default 'leecher' category is configured this way out of the box:

* One file at a time for anyone identified as a 'leecher', with the queue strategy set to 'Round robin'
* Download speed capped at 100kb/s
* Lowest priority (all other downloads from all other users must complete first)

I still need to implement quotas (#388) to control how many files users can enqueue at a given time, and when I do I will most likely default the 'leecher' category to just 1 enqueued file.

With all of these controls in place, I really can't think of a good faith argument in support of locked files; all of the excuses about bandwidth or clogging queues are negated, and the only left are those about creating artificial scarcity, or, in the worst cases, forcing people to pay for access.

All this being said, I'm definitely open to discussion, and I'd like to hear any rationale for locked files that I haven't discussed already.

**Reply by jpdillingham** [reply]
With #388 complete, users can now control how many files and/or megabytes other users can enqueue or download on a daily or weekly basis.

What this means is that users for whom people that want to incentivize sharing would typically lock files for (e.g. "leechers") can be restricted to:

* Downloading 1 file at a time, at 100kb/s, and at the lowest priority
* Enqueuing 1 file at a time
* Downloading 1 file per day, and 1 file per week

This is an ideal alternative to locking files, as it creates a huge incentive to share that has direct impact (share more -> move out of the "leecher" group automatically -> download more) while offering just the tiniest amount of grace to people who may not be sharing for completely innocent reasons.

**Comment by justjakka** [comment]
well for me locked folders are crucial, since i cant be bothered adding thousands of folders manually, seeing if i can share them or not. hence id just made my library searchable but not downloadable, so people would pm me for files and id send them if i can.

**Reply by jpdillingham** [reply]
Why wouldn't you be able to share them?

**Reply by justjakka** [reply]
releases that i got from artists personally, with a request to keep the files only to myself, my and my friends releases (examples for music since i share only it)

**Reply by jpdillingham** [reply]
Would adding a [share exclusion](https://github.com/slskd/slskd/blob/master/docs/config.md#shares) (adding a directory with a `!` in front of the name will prevent the directory from being shared) be enough?  Alternatively you could add a [filter](https://github.com/slskd/slskd/blob/master/docs/config.md#filters) to exclude artist or track names regardless of which directory they are in.

**Reply by justjakka** [reply]
directories starting with ! wouldnt get added and applying a filter wouldnt help anyways, my library is too big to remember everything so i have to manually exclude folders, which would take god knows how many hours

**Comment by wovels** [comment]
absolutely necessary feature and the only reason why I'm not switching to slskd

**Reply by jpdillingham** [reply]
Can you elaborate on how you would use this?

**Reply by wovels** [reply]
I have my locked files accessible only to fellow vinyl enthusiasts of music who understand and value the effort that is put into making a collection. Or, if someone doesn't have a shared folder (or can't share because they're on the road a lot) they can find me this way and we can exchange in a fair manner (vinyl EP for vinyl EP).

This allows for fair exchange that benefits both sides. 

So much effort and money goes into finding a record, cleaning it, making a decent rip, manually de-clicking it for hours, formatting files, tags, scanning cover, etc.

Personally, I see nothing wrong with such fair exchange.

**Reply by wovels** [reply]
PS: open share with 1 slot and 100kb/s still allows people to leach easily. It just might take a few days to leach an EP but it is still possible. Besides, nothing stops them from making as many new Soulseek user account as it takes to leach it all.

**Reply by jpdillingham** [reply]
What if you could set a quota per username/IP address to prevent leechers from enqueueing more than 1 file at a time, or downloading more than 1 file a day/week/month?

**Reply by wovels** [reply]
It's too easy to make a new username (Login as>{something random like vhf7d8vfd78})

Ability to open for trusted group / lock for everyone else would do just fine. Displaying locked files helps finding like-minded people (soul seek).

**Comment by thomask** [comment]
Same for me for exactly the same reasons as @justjakka and @vvkanz have outlined. Perhaps the official client having this feature counts for something? 

In any case, I would be more than happy to try and prepare a PR. Although that would mean learning some C# and the Soulseek API 😅

In any case, many thanks for your consideration!

Edit: Re-read your message above and edited accordingly

**Reply by jpdillingham** [reply]
I don't want to add it because I believe it is harmful to the Soulseek network.  It's also harmful to any application that aims to automate anything -- like slskd -- because it's adding a manual step to the process.

I'm not yet convinced that any good faith use case for locking files is unable to be mitigated through other means.  Quotas and rate limits remove the objections over download quantity, up to and including limiting the ability of anyone that hasn't been explicitly added to a list to downloading 1 file per month at a rate of 1kb/s (theoretically, this hasn't been implemented, but it could).

The rest of the motivation seems to come down to users simply wanting to advertise files while retaining the ability to make a case-by-case determination regarding whether to grant access, and while I can understand people that want to do this to force people to share, or to share some arbitrary quantity, quality, or genre of files, it also enables people to restrict access in trade-only scenarios, or where the uploader is selling access.

In short, locked files enable abusive and bad faith uploaders, and unless I hear a really convincing argument for a good-faith use case, I'm not really motivated to spend my time adding the feature (or helping anyone add it).

**Comment by StayBlue** [comment]
I would like to start by stating that I understand not wanting to support something along ideological lines, especially as this is an open-source project. I would like to provide an argument as to why this might ultimately be a good thing.

I am not very much involved in Soulseek, but I like its idea. You share files with others, and you are free to download other people's files as well. However, there are some cases, regardless of how we personally view the reason, that someone might want to make certain files or folders inaccessible. I do think pay walling certain files or folders is scummy, but at the end of the day, sometimes people pay for that media and would prefer to get some compensation for sharing it. You could argue that the option to get the file or folder from some source, whereas you couldn't find it anywhere else (if that wasn't an option), is better than nothing at all.

I believe giving control to the users is important, even if it means making other people's lives harder. In my case, the fact that I can't make it so that one directory is locked means that I can't share the thousands of songs in my music directory altogether, and I'm sure others feel the same way. I think a middle ground would be to make it so you can put certain directories in a blacklist so they just aren't propagated onto Soulseek. This way, people wouldn't have to put files they don't want to share in a completely different directory just so they don't have to share it, which they can do now, but this option would make it so they don't have to go out of their way to do it. It would also prevent people from putting paywalls on content since it wouldn't show the content to begin with.

**Reply by SyntaxJunkie** [reply]
You can already exclude nested directories with [share exclusions](https://github.com/slskd/slskd/blob/master/docs/config.md#shares), or even supply a [regex filter](https://github.com/slskd/slskd/blob/master/docs/config.md#filters) if you wanted to.
If you want to charge for content that isn't yours in the first place you can go somewhere else.

**Reply by StayBlue** [reply]
If the filters feature works for folders too, then that should be made more clear in the documentation as I was under the impression that you had to exclude each file individually.

It would be nice if these exclusions could be done on the dashboard itself for convenience's sake.

**Reply by SyntaxJunkie** [reply]
> If the filters feature works for folders too, then that should be made more clear in the documentation as I was under the impression that you had to exclude each file individually.
> 
> It would be nice if these exclusions could be done on the dashboard itself for convenience's sake.

That's fair. Again for your use case, the share exclusions would likely better suit you. But if you'd need a regex, it seems filters are applied to directories as well as files, after reviewing [ShareScanner.cs](https://github.com/slskd/slskd/blob/master/src/slskd/Shares/ShareScanner.cs)

**Comment by HVR88** [comment]
necro post, but needed to be said:

>someone might want to make certain files or folders inaccessible.

Then don't share them.  IMO, the lock needs to be removed from all clients, including the official. If someone doesn't like it, they can take the rest of their files and hit the bricks too. No need for nuance here, this issue is black and white. Don't share what you don't want to share.

I'm with the author here and glad this client doesn't offer that feature. It's exclusionary.  My shares are much larger than most, almost all FLAC, and I offer over 70MB/s speeds to downloaders. Yet I still get automated messages from elitist idiots  who don't know how to set up their client, complaining that I'm not sharing. DOH!
